homeassistant/components/conversation/default_agent.py

```diff
@@ -50,93 +43,178 @@ def async_register(hass, intent_type, utterances):
         )
 
-        @callback
-        def component_loaded(event):
-            """Handle a new component loaded."""
-            self.register_utterances(event.data[ATTR_COMPONENT])
+    def get_or_load_intents(self, language: str) -> LanguageIntents | None:
+        """Load all intents for language."""
+        lang_intents = self._lang_intents.get(language)
 
-        self.hass.bus.async_listen(EVENT_COMPONENT_LOADED, component_loaded)
+        if lang_intents is None:
+            intents_dict: dict[str, Any] = {}
+            loaded_components: set[str] = set()
+        else:
+            intents_dict = lang_intents.intents_dict
+            loaded_components = lang_intents.loaded_components
 
-        # Check already loaded components.
+        # Check if any new components have been loaded
+        intents_changed = False
         for component in self.hass.config.components:
-            self.register_utterances(component)
+            if component in loaded_components:
+                continue
 
-    @callback
-    def register_utterances(self, component):
-        """Register utterances for a component."""
-        if component not in UTTERANCES:
-            return
-        for intent_type, sentences in UTTERANCES[component].items():
-            async_register(self.hass, intent_type, sentences)
+            # Don't check component again
+            loaded_components.add(component)
 
-    async def async_process(
-        self,
-        text: str,
-        context: core.Context,
-        conversation_id: str | None = None,
-        language: str | None = None,
-    ) -> ConversationResult | None:
-        """Process a sentence."""
-        intents = self.hass.data[DOMAIN]
+            # Check for intents for this component with the target language
+            component_intents = get_intents(component, language)
+            if component_intents:
+                # Merge sentences into existing dictionary
+                merge_dict(intents_dict, component_intents)
+
+                # Will need to recreate graph
+                intents_changed = True
+
+        if not intents_dict:
+            return None
+
+        if not intents_changed and (lang_intents is not None):
+            return lang_intents
 
-        for intent_type, matchers in intents.items():
-            for matcher in matchers:
-                if not (match := matcher.match(text)):
+        # This can be made faster by not re-parsing existing sentences.
+        # But it will likely only be called once anyways, unless new
+        # components with sentences are often being loaded.
+        intents = Intents.from_dict(intents_dict)
+
+        if lang_intents is None:
+            lang_intents = LanguageIntents(intents, intents_dict, loaded_components)
+            self._lang_intents[language] = lang_intents
+        else:
+            lang_intents.intents = intents
+
+        return lang_intents
```

Comment:
Since this method checks `self._lang_intents` at the beginning, I would have expected it to also store the result here?

Reply:
Ah it does just above it.